### ISSUE TITLE
Cannot make NEST 2.16.0 - cannot find GSL header files

### ISSUE THREAD

**Issue by jeliason** [nest/nest-simulator#1044]
Hi!

I've been trying for some time to install NEST 2.16.0 on Ubuntu 16.04 to run with Python and PyNN. I want to have the GSL libraries linked. The configuration with cmake yields the following:

cmake -DCMAKE_INSTALL_PREFIX:PATH=/home/myname/opt/nest-simulator-2.16.0-build /home/myname/opt/nest-simulator-2.16.0

-- The CXX compiler identification is GNU 7.3.0
-- The C compiler identification is GNU 7.3.0
-- Check for working CXX compiler: /home/myname/anaconda2/envs/data_env3/bin/x86_64-conda_cos6-linux-gnu-c++
-- Check for working CXX compiler: /home/myname/anaconda2/envs/data_env3/bin/x86_64-conda_cos6-linux-gnu-c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Check for working C compiler: /home/myname/anaconda2/envs/data_env3/bin/x86_64-conda_cos6-linux-gnu-cc
-- Check for working C compiler: /home/myname/anaconda2/envs/data_env3/bin/x86_64-conda_cos6-linux-gnu-cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Could NOT find Doxygen (missing:  DOXYGEN_EXECUTABLE) 
-- Looking for include file inttypes.h
-- Looking for include file inttypes.h - found
-- Looking for include file mach-o/dyld.h
-- Looking for include file mach-o/dyld.h - not found
-- Looking for include file mach/mach.h
-- Looking for include file mach/mach.h - not found
-- Looking for include file memory.h
-- Looking for include file memory.h - found
-- Looking for include file stdint.h
-- Looking for include file stdint.h - found
-- Looking for include file sys/types.h
-- Looking for include file sys/types.h - found
-- Looking for C++ include istream
-- Looking for C++ include istream - found
-- Looking for C++ include ostream
-- Looking for C++ include ostream - found
-- Looking for C++ include sstream
-- Looking for C++ include sstream - found
-- Looking for stddef.h
-- Looking for stddef.h - found
-- Check size of long long
-- Check size of long long - done
-- Check size of u_int16_t
-- Check size of u_int16_t - failed
-- Check size of uint16_t
-- Check size of uint16_t - done
-- Looking for NAN
-- Looking for NAN - found
-- Looking for isnan
-- Looking for isnan - found
-- Looking for M_E
-- Looking for M_E - found
-- Looking for M_PI
-- Looking for M_PI - found
-- Looking for expm1
-- Looking for expm1 - not found
-- Host triple: x86_64-pc-linux
-- Target triple: x86_64-pc-linux
-- Found LTDL: /home/myname/anaconda2/lib/libltdl.so (found version "2.4.6") 
-- Could NOT find Readline (missing:  READLINE_LIBRARY) (found version "6.3")
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29.1") 
-- Found GSL: /usr/include (found version "2.1") 
-- Found PythonInterp: /home/myname/anaconda2/envs/data_env3/bin/python (found version "3.6.6") 
-- Found PythonLibs: /home/myname/anaconda2/envs/data_env3/lib/libpython3.6m.so (found suitable exact version "3.6.6") 
-- Found Cython: /home/myname/anaconda2/envs/data_env3/bin/cython (found version "0.28.5") 
-- Try OpenMP C flag = [-fopenmp]
-- Performing Test OpenMP_FLAG_DETECTED
-- Performing Test OpenMP_FLAG_DETECTED - Success
-- Try OpenMP CXX flag = [-fopenmp]
-- Performing Test OpenMP_FLAG_DETECTED
-- Performing Test OpenMP_FLAG_DETECTED - Success
-- Found OpenMP: -fopenmp  
-- Check the abort exitcode.
-- Check the abort exitcode. 134
-- Check the segmentation fault exitcode.
-- Check the segmentation fault exitcode. 139
-- Check whether the compiler ignores cmath makros.
-- Check whether the compiler ignores cmath makros. OFF
-- Check whether the compiler does NOT include <*.h> headers ISO conformant.
-- Check whether the compiler does NOT include <*.h> headers ISO conformant. OFF
-- Check whether the compiler respects symbolic signal names in signal.h.
-- Check whether the compiler respects symbolic signal names in signal.h. OFF
-- Check static template member declaration.
-- Check static template member declaration. OFF
-- Check for STL vector capacity base unity.
-- Check for STL vector capacity base unity. ON
-- Check for STL vector capacity doubling strategy.
-- Check for STL vector capacity doubling strategy. ON
-- Check whether the compiler fails with ICE.
-- Check whether the compiler fails with ICE. OFF
-- Check if ::nan is available from cmath.
-- Check if ::nan is available from cmath. ON
-- Check if ::isnan is available from cmath.
-- Check if ::isnan is available from cmath. ON
-- Found PythonLibs: /home/myname/anaconda2/envs/data_env3/lib/libpython3.6m.so (found version "3.6.6") 

--------------------------------------------------------------------------------
NEST Configuration Summary
--------------------------------------------------------------------------------

Build type          : 
Target System       : Linux
Cross Compiling     : FALSE
C compiler          : /home/myname/anaconda2/envs/data_env3/bin/x86_64-conda_cos6-linux-gnu-cc
C compiler flags    : -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -pipe  -O2 -Wall -fopenmp -fdiagnostics-color=auto
C++ compiler        : /home/myname/anaconda2/envs/data_env3/bin/x86_64-conda_cos6-linux-gnu-c++
C++ compiler flags  : -fvisibility-inlines-hidden -std=c++17 -fmessage-length=0 -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -pipe  -std=c++11 -O2 -Wall -fopenmp -fdiagnostics-color=auto
Build dynamic       : ON

Built-in modules    : models;precise;topology
User modules        : None
Python bindings     : Yes (Python 3.6.6: /home/myname/anaconda2/envs/data_env3/bin/python)
       Includes     : /home/myname/anaconda2/envs/data_env3/include/python3.6m
       Libraries    : /home/myname/anaconda2/envs/data_env3/lib/libpython3.6m.so

Cython bindings     : Yes (Cython 0.28.5: /home/myname/anaconda2/envs/data_env3/bin/cython)
Use threading       : Yes (OpenMP: -fopenmp)
Use GSL             : Yes (GSL 2.1)
    Includes        : /usr/include
    Libraries       : /usr/lib/x86_64-linux-gnu/libgsl.so;/usr/lib/x86_64-linux-gnu/libgslcblas.so

Use Readline        : No
Use libltdl         : Yes (LTDL 2.4.6)
    Includes        : /home/myname/anaconda2/include
    Libraries       : /home/myname/anaconda2/lib/libltdl.so

Use doxygen         : No
Use MPI             : No
Use MUSIC           : No
Use libneurosim     : No
Use Boost           : No

--------------------------------------------------------------------------------

The NEST executable will be installed to:
  /home/myname/opt/nest-simulator-2.16.0-build/bin/

Documentation and examples will be installed to:
  /home/myname/opt/nest-simulator-2.16.0-build/share/doc/nest/

PyNEST will be installed to:
    /home/myname/opt/nest-simulator-2.16.0-build/lib/python3.6/site-packages

To set necessary environment variables, add the following line
to your ~/.bashrc :
  source /home/myname/opt/nest-simulator-2.16.0-build/bin/nest_vars.sh

--------------------------------------------------------------------------------

You can now build and install NEST with
  make
  make install
  make installcheck

If you experience problems with the installation or the use of NEST,
please see http://www.nest-simulator.org/frequently_asked_questions
or go to http://www.nest-simulator.org/community to find out how to
join the user mailing list.

-- Configuring done
-- Generating done
-- Build files have been written to: /home/myname/opt/nest-simulator-2.16.0-build

### Make command
Following this with the make command yields:

make
Scanning dependencies of target nestutil
[  1%] Building CXX object libnestutil/CMakeFiles/nestutil.dir/stopwatch.cpp.o
[  1%] Building CXX object libnestutil/CMakeFiles/nestutil.dir/numerics.cpp.o
/home/myname/opt/nest-simulator-2.16.0/libnestutil/numerics.cpp:63:10: fatal error: gsl/gsl_math.h: No such file or directory
 #include <gsl/gsl_math.h>
          ^~~~~~~~~~~~~~~~
compilation terminated.
libnestutil/CMakeFiles/nestutil.dir/build.make:86: recipe for target 'libnestutil/CMakeFiles/nestutil.dir/numerics.cpp.o' failed
make[2]: *** [libnestutil/CMakeFiles/nestutil.dir/numerics.cpp.o] Error 1
CMakeFiles/Makefile2:336: recipe for target 'libnestutil/CMakeFiles/nestutil.dir/all' failed
make[1]: *** [libnestutil/CMakeFiles/nestutil.dir/all] Error 2
Makefile:160: recipe for target 'all' failed
make: *** [all] Error 2

This doesn't make sense, as I have libgsl-dev installed (and tried reinstalling as well), and the gsl header files are in their proper place in /usr/include/gsl. I have all of the required libraries and development packages installed, as per the NEST installation page. I'm not sure what to change - compiling a random script that links to the header files using gcc seems to work just fine.

Any hints would be most appreciated.

**Comment by heplesser** [comment]
Your problems may be caused by confusion caused by Anaconda. We have experienced problems caused by Anaconda earlier and will take a look.

@alexvanmeegen could you take a look?

**Comment by clinssen** [comment]
Relevant discussion: https://gitlab.kitware.com/cmake/cmake/issues/17966

It looks like CMake explicitly excludes "/usr/include" from the include path, because it is detected as the compiler's standard search path. However, for reasons of virtual environment isolation, Anaconda explicitly *excludes* it from the path. @jeliason: to check if this is indeed the issue, could you try the workaround described on the above URL (i.e. add the directory explicitly to `target_include_directories`)?

**Comment by jeliason** [comment]
I actually just got it to compile by having GSL installed by conda in the conda env, then linking to those local GSL header and library files with compiler flags **with-gsl** and **GSL_INCLUDE_DIR** during configuration. I'm not sure why this local one had to be used, since I added the system directories explicitly previously and that didn't seem to work. I got through make and make install, and importing nest in the python interpreter causes no issue, so hopefully there will be no problems with PyNEST in the future.
